File: py/matrix.py
# Matrix class in Python
epsilon = 1e-6
def correct_float_error(f):
    # Takes a float and rounds to integer if it is close enough
    if abs(f - round(f)) < epsilon:
        return round(f)
    else:
        return f

class Matrix:
    def __init__(self, rows, cols):
        self.rows = rows
        self.cols = cols
        self.data = [[0 for i in range(cols)] for j in range(rows)]

    def __add__(self, other):
        if self.rows != other.rows or self.cols != other.cols:
            raise Exception("Matrices must have the same dimensions")
        result = Matrix(self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                result.data[i][j] = self.data[i][j] + other.data[i][j]
        return result

    def __sub__(self, other):
        if self.rows != other.rows or self.cols != other.cols:
            raise Exception("Matrices must have the same dimensions")
        result = Matrix(self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                result.data[i][j] = self.data[i][j] - other.data[i][j]
        return result

    def __mul__(self, other):
        if self.cols != other.rows:
            raise Exception("Matrices must have the same dimensions")
        result = Matrix(self.rows, other.cols)
        for i in range(self.rows):
            for j in range(other.cols):
                for k in range(self.cols):
                    result.data[i][j] += self.data[i][k] * other.data[k][j]
        return result

    def __truediv__(self, other):
        if self.rows != other.rows or self.cols != other.cols:
            raise Exception("Matrices must have the same dimensions")
        result = Matrix(self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                result.data[i][j] = self.data[i][j] / other.data[i][j]
        return result

    def __pow__(self, other):
        if self.rows != self.cols:
            raise Exception("Matrix must be square")
        result = Matrix(self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                result.data[i][j] = self.data[i][j] ** other
        return result

    def __mod__(self, other):
        if self.rows != self.cols:
            raise Exception("Matrix must be square")
        result = Matrix(self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                result.data[i][j] = self.data[i][j] % other
        return result

    def __eq__(self, other):
        if self.rows != other.rows or self.cols != other.cols:
            return False
        for i in range(self.rows):
            for j in range(self.cols):
                if self.data[i][j] != other.data[i][j]:
                    return False
        return True

    def __ne__(self, other):
        return not self == other

    def __getitem__(self, key):
        return self.data[key]

    def __setitem__(self, key, value):
        self.data[key] = value

    def __len__(self):
        return self.rows

    def __iter__(self):
        return iter(self.data)

    def get(self, row, col):
        return self.data[row][col]

    def set(self, row, col, value):
        self.data[row][col] = value

    def transpose(self):
        result = Matrix(self.cols, self.rows)
        for i in range(self.rows):
            for j in range(self.cols):
                result.data[j][i] = self.data[i][j]
        return result

    def map(self, func):
        for i in range(self.rows):
            for j in range(self.cols):
                self.data[i][j] = func(self.data[i][j])

    def copy(self):
        result = Matrix(self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                result.data[i][j] = self.data[i][j]
        return result

    def randomize(self):
        import random
        for i in range(self.rows):
            for j in range(self.cols):
                self.data[i][j] = random.random()

    def print(self):
        # Create a lines list with num columns + 2 slots
        lines = ["" for i in range(self.cols + 2)]
        lines[0] = "┌\t"
        for i in range(self.cols):
            lines[0] += "\t"
        lines[0] += "┐"
        for i in range(self.rows):
            lines[i+1] = "│"
            for j in range(self.cols):
                lines[i+1] += "\t" + str(correct_float_error(self.data[i][j]))
            lines[i+1] += "\t│"
        lines[self.rows+1] = "└\t"
        for i in range(self.cols):
            lines[self.rows+1] += "\t"
        lines[self.rows+1] += "┘"
        return "\n".join(lines)

    def setData(self, data):
        self.data = data

    def __str__(self):
        return self.print()

    def __repr__(self):
        return self.print()

File: js/matrix.js
// Matrix class for Row Echelon Form
class Matrix {
	constructor(rows, cols) {
		this.rows = rows;
		this.cols = cols;
		this.data = [];
		for (let i = 0; i < this.rows; i++) {
			this.data[i] = [];
			for (let j = 0; j < this.cols; j++) {
				this.data[i][j] = 0;
			}
		}
	}

	static fromArray(arr) {
		let m = new Matrix(arr.length, 1);
		for (let i = 0; i < arr.length; i++) {
			m.data[i][0] = arr[i];
		}
		return m;
	}

	toArray() {
		let arr = [];
		for (let i = 0; i < this.rows; i++) {
			for (let j = 0; j < this.cols; j++) {
				arr.push(this.data[i][j]);
			}
		}
		return arr;
	}

	randomize() {
		for (let i = 0; i < this.rows; i++) {
			for (let j = 0; j < this.cols; j++) {
				this.data[i][j] = Math.floor(Math.random() * 10);
			}
		}
	}

	add(n) {
		if (n instanceof Matrix) {
			for (let i = 0; i < this.rows; i++) {
				for (let j = 0; j < this.cols; j++) {
					this.data[i][j] += n.data[i][j];
				}
			}
		} else {
			for (let i = 0; i < this.rows; i++) {
				for (let j = 0; j < this.cols; j++) {
					this.data[i][j] += n;
				}
			}
		}
	}

	static subtract(a, b) {
		let result = new Matrix(a.rows, a.cols);
		for (let i = 0; i < result.rows; i++) {
			for (let j = 0; j < result.cols; j++) {
				result.data[i][j] = a.data[i][j] - b.data[i][j];
			}
		}
		return result;
	}

	static multiply(a, b) {
		if (a.cols !== b.rows) {
			console.log('Columns of A must match rows of B.');
			return undefined;
		}
		let result = new Matrix(a.rows, b.cols);
		for (let i = 0; i < result.rows; i++) {
			for (let j = 0; j < result.cols; j++) {
				let sum = 0;
				for (let k = 0; k < a.cols; k++) {
					sum += a.data[i][k] * b.data[k][j];
				}
				result.data[i][j] = sum;
			}
		}
		return result;
	}

	static transpose(matrix) {
		let result = new Matrix(matrix.cols, matrix.rows);
		for (let i = 0; i < matrix.rows; i++) {
			for (let j = 0; j < matrix.cols; j++) {
				result.data[j][i] = matrix.data[i][j];
			}
		}
		return result;
	}

	static map(matrix, func) {
		let result = new Matrix(matrix.rows, matrix.cols);
		for (let i = 0; i < matrix.rows; i++) {
			for (let j = 0; j < matrix.cols; j++) {
				let val = matrix.data[i][j];
				result.data[i][j] = func(val);
			}
		}
		return result;
	}

	static map2(a, b, func) {
		let result = new Matrix(a.rows, a.cols);
		for (let i = 0; i < a.rows; i++) {
			for (let j = 0; j < a.cols; j++) {
				let x = a.data[i][j];
				let y = b.data[i][j];
				result.data[i][j] = func(x, y);
			}
		}
		return result;
	}

	print() {
		console.table(this.data);
	}

	serialize() {
		return JSON.stringify(this);
	}

	static deserialize(data) {
		if (typeof data == 'string') {
			data = JSON.parse(data);
		}
		let matrix = new Matrix(data.rows, data.cols);
		matrix.data = data.data;
		return matrix;
	}

	copy() {
		let matrix = new Matrix(this.rows, this.cols);
		matrix.data = this.data;
		return matrix;
	}

	static fromArray(arr) {
		return new Matrix(arr.length, 1, arr);
	}

	toArray() {
		let arr = [];
		for (let i = 0; i < this.rows; i++) {
			for (let j = 0; j < this.cols; j++) {
				arr.push(this.data[i][j]);
			}
		}
		return arr;
	}

	toString() {
		let str = '';
		for (let i = 0; i < this.rows; i++) {
			let row;
			if (this.rows === 1) {
				row = "["
			} else if (i === 0) {
				row = '┌';
			} else if (i === this.rows - 1) {
				row = '└';
			} else {
				row = '│';
			}
			for (let j = 0; j < this.cols; j++) {
				row += this.data[i][j];
				if (j < this.cols - 1) {
					row += ' ';
				}
			}
			if (this.rows === 1) {
				row += ']';
			} else if (i === 0) {
				row += '┐';
			} else if (i === this.rows - 1) {
				row += '┘';
			} else {
				row += '│';
			}
			str += row;
			if (i < this.rows - 1) {
				str += '\n';
			}
		}
		return str;
	}

	get(x, y) {
		return this.data[x][y];
	}

	set(x, y, value) {
		this.data[x][y] = value;
	}

	setRow(row, arr) {
		this.data[row] = arr;
	}

	setCol(col, arr) {
		for (let i = 0; i < this.rows; i++) {
			this.data[i][col] = arr[i];
		}
	}

	getRow(row) {
		return this.data[row];
	}

	getCol(col) {
		let arr = [];
		for (let i = 0; i < this.rows; i++) {
			arr.push(this.data[i][col]);
		}
		return arr;
	}

	setAll(value) {
		for (let i = 0; i < this.rows; i++) {
			for (let j = 0; j < this.cols; j++) {
				this.data[i][j] = value;
			}
		}
	}

	randomize() {
		for (let i = 0; i < this.rows; i++) {
			for (let j = 0; j < this.cols; j++) {
				this.data[i][j] = Math.random() * 2 - 1;
			}
		}
	}
}



module.exports = Matrix;